presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.spi.PrestoException;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+
+public class HiveBucketProperty
+{
+    private final List<String> clusteredBy;
+    private final int bucketCount;
+
+    @JsonCreator
+    public HiveBucketProperty(@JsonProperty("clusteredBy") List<String> clusteredBy, @JsonProperty("bucketCount") int bucketCount)
+    {
+        this.clusteredBy = clusteredBy;
+        this.bucketCount = bucketCount;
+    }
+
+    public static Optional<HiveBucketProperty> fromStorageDescriptor(StorageDescriptor storageDescriptor, String tablePartitionName)
+    {
+        boolean bucketColsSet = storageDescriptor.isSetBucketCols() && !storageDescriptor.getBucketCols().isEmpty();
+        boolean numBucketsSet = storageDescriptor.isSetNumBuckets() && storageDescriptor.getNumBuckets() > 0;
+        if (!bucketColsSet || !numBucketsSet) {
+            if (bucketColsSet || numBucketsSet) {
```

Comment:
I would put this first as `if (bucketColsSet != numBucketsSet)` and the the remain `if` structure is much simpler